whylog/constraints/__init__.py

```diff
@@ -2,42 +2,184 @@
+        :param groups: all groups that are linked by constraint,
+                       represented by list of tuples (line_id, group_no),
+                       where line_id and group_no is inner numeration between Front and Teacher
+        """
+        self.params = param_dict
+        self.groups = groups
+
+    @abstractmethod
+    def convert_to_user_constraint_intent(self):
+        """
+        Converts constraint to UserConstraintIntent object.
+
+        For Teacher and Config use while saving constraint into Whylog knowledge base.
+        """
+        return UserConstraintIntent(self.TYPE, self.groups, self.params)
+
+    @classmethod
+    def get_param_names(cls):
+        """
+        Returns names of constraint additional params.
+
+        For Front to display param names to user and then ask user for param contents.
+        """
+        raise NotImplementedError("Subclass should implement this")
+
+    @classmethod
+    def get_groups_count(cls):
+        """
+        Returns minimal and maximal count of groups needed to create constraint.
+
+        For Front to ask user for proper count of groups.
+        2, None - at least 2 groups
+        2, 2 - exactly 2 groups
+        """
+        return cls.MIN_GROUPS_COUNT, cls.MAX_GROUPS_COUNT
+
+    @classmethod
+    def verify(cls, param_dict, group_contents):
+        """
+        Verifies constraint for given params in param_dict and groups contents.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of groups contents,
+
+        For LogReader and Teacher verification.
+        It must be optimized as well as possible (for LogReader).
+        """
+        raise NotImplementedError("Subclass should implement this")
 
 
 class TimeConstraint(AbstractConstraint):
-    def __init__(self, line_earlier, line_later, min_delta=None, max_delta=None):
-        pass
+    """
+    Time delta between two dates must be greater than 'max_delta' and lower than 'min_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = 2
+
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    def __init__(self, param_dict, groups):
+        """
+        :param groups: First element of groups is a group with earlier date.
+
+        I.e:
+        TimeConstraint(
+            {'min_delta': 12, 'max_delta' 30},
+            [(0, 1), (2, 1)]
+        )
+        """
+        super(TimeConstraint, self).__init__(param_dict, groups)
+        self._check_params()
 
+    def _check_params(self):
+        param_names = self.params.keys()
+        if self.MIN_DELTA not in param_names and self.MAX_DELTA not in param_names:
+            raise ConstructorParamsError(self.TYPE, self.get_param_names(), param_names)
+        if not len(self.groups) == 2:
+            raise ConstructorGroupsError(
+                self.TYPE, len(self.groups), self.MIN_GROUPS_COUNT, self.MAX_GROUPS_COUNT
```

Comment:
MIN_GROUPS_COUNT and MAX_GROUPS_COUNT doesn't make sens in this class. You can make _check_groups function in Abstract constraint and do not copy code that should be the same in all inherited classes. You can make the same mechanism for checking param in paramdict by using PARAMS attribute.